framework/src/preconditioners/DualMortarPreconditioner.C

```diff
@@ -0,0 +1,588 @@
+          _dofmap->dof_indices(node_bdry, di, vn);
+          for (auto it = di.begin(); it!=di.end(); ++it)
+            if(std::find(_dof_sets_secondary[vn].begin(), _dof_sets_secondary[vn].end(), *it)==_dof_sets_secondary[vn].end())
+              _dof_sets_secondary[vn].push_back(*it);
+      }
+
+      if (boundary_id == _primary_boundary)
+      {
+        _dofmap->dof_indices(node_bdry, di, vn);
+        for (auto it = di.begin(); it!=di.end(); ++it)
+          if(std::find(_dof_sets_primary[vn].begin(), _dof_sets_primary[vn].end(), *it)==_dof_sets_primary[vn].end())
+            _dof_sets_primary[vn].push_back(*it);
+      }
+    }
+  }
+}
+
+void
+DualMortarPreconditioner::printDofSets()
+{
+  /// Dofs in the interior
+  for (auto it1 = _dof_sets_interior.begin(); it1 != _dof_sets_interior.end(); it1++)
+  {
+    std::cout << "variable "<<std::endl;
+    for (auto it2 = it1->begin(); it2 != it1->end(); it2++)
+    {
+      std::cout << " \tdofs in subdomain " << it2->first<<": ";
+      for (auto it3 = it2->second.begin(); it3 != it2->second.end(); it3++)
+        std::cout <<*it3 << " ";
+      std::cout <<"\n";
+    }
+  }
+
+  // primary
+  std::cout<<"primary dofs:\n";
+  for (auto it1=_dof_sets_primary.begin(); it1!=_dof_sets_primary.end(); it1++)
+  {
+    std::cout<< "\tvariable :";
+    for (auto it2=it1->begin(); it2!=it1->end(); it2++)
+      std::cout<<*it2<<" ";
+    std::cout<<std::endl;
+  }
+
+  // secondary
+  std::cout<<"secondary dofs:\n";
+  for (auto it1=_dof_sets_secondary.begin(); it1!=_dof_sets_secondary.end(); it1++)
+  {
+    std::cout<< "\tvariable :";
+    for (auto it2=it1->begin(); it2!=it1->end(); it2++)
+      std::cout<<*it2<<" ";
+    std::cout<<std::endl;
+  }
+}
+
+void
+DualMortarPreconditioner::condenseSystem()
+{
+  std::vector<dof_id_type> u1c=_dof_sets_primary[0];
+  std::vector<dof_id_type> u2c=_dof_sets_secondary[0];
+
+  std::vector<dof_id_type> lm=_dof_sets_secondary[1];
+
+  std::vector<dof_id_type> u1i=_dof_sets_interior[0][_primary_subdomain];
+  std::vector<dof_id_type> u2i=_dof_sets_interior[0][_secondary_subdomain];
+
+  _matrix->create_submatrix(* _D, lm, u2c);
+  _D->get_transpose(*_D);
+  _matrix->create_submatrix(* _M, lm, u1c);
+  _matrix->create_submatrix(* _MDinv, u1c, lm);
+  _M->get_transpose(*_M);
+  _M->close();
+
+  _matrix->create_submatrix(* _K2ci, u2c, u2i);
+  _matrix->create_submatrix(* _K2cc, u2c, u2c);
+
+  _K2ci->close();
+  _K2cc->close();
+
+  // invert _D:
+  // _D should be strictly diagonal if dual_mortar approach is utilized
+  // so we only need to compute the reciprocal number of the diagonal entries
+  // to save memory, no new matrix is created
+  for (unsigned int i=0; i<_D->m(); ++i)
+  {
+   Number tmp = 1.0/(*_D)(i,i);
+   std::vector<numeric_index_type> row_i;
+   row_i.push_back(i);
+    _D->zero_rows(row_i, tmp);
+  }
+  _D->close();
+
+
+  // compute MDinv=_M*_D
+  _M->matrix_matrix_mult(*_D, *_MDinv); // (should use empty initializer for _MDinv)
```

Comment:
We could simply scale _M to get MD since D is a vector

Reply:
I agree it would be better to do the scale! However, I do not see such a function in the class. Is there?